### PR TITLE
Embeddings: add 'model version' to indexes

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR updates the embeddings indexes to include the model used to generate the indexes. With this information, we now
* Throw a search error if the current embeddings config does not match the model used to generate the index. This avoids bad situations like a dimension mismatch between query and index vectors, or silently returning poor results.
* When reembedding a repo, always fall back to a full index if the embeddings model has changed.

To support this, the PR adds the concept of "index format version" to embeddings indexes. This lets us start to encode the embeddings model, while still handling old indexes. This is generally useful to have, as it lets us add other metadata or change the index format in the future.

Addresses #52668

## Test plan

New unit tests: mismatched query and index models, encoding and decoding index format version.

Manual testing
* Checked that search throws error when config doesn't match index model
* Checked we can still decode older indexes

I'm not happy with the level of automated testing in the PR, especially for reindexing (checking when we fall back to a full index). Adding these tests requires a refactor, which I will tackle in a follow-up.

### CODE REVIEW

**Comment by jtibshirani** [comment]
I took a classic "self-describing files" approach and decided to add this metadata to the index files themselves. I could have added it to the jobs table instead, but it wasn't much more convenient... and that table is really supposed to be about jobs, it didn't feel like the right place to be adding index metadata. In the future, we may want a new table that describes the *current state* of all repo embeddings, which would include metadata like this.

I'm super happy to hear different opinions here, let me know what you think about the approach :)

File: enterprise/cmd/embeddings/shared/query_embeddings_cache.go [hunk 1]
```diff
@@ -19,20 +19,21 @@ func getCachedQueryEmbeddingFn() (getQueryEmbeddingFn, error) {
 		return nil, errors.Wrap(err, "creating query embeddings cache")
 	}
 
-	return func(ctx context.Context, query string) (queryEmbedding []float32, err error) {
+	return func(ctx context.Context, query string) (queryEmbedding []float32, model string, err error) {
```
Comment (jtibshirani) [inline enterprise/cmd/embeddings/shared/query_embeddings_cache.go:22]: Returning the model here is not very elegant. In a follow-up, I want to consider removing this caching altogether, which will make this cleaner. The reasoning:
* Caching query embeddings no longer seems super, now that we don't have the context detection index. Before, it was helpful because we'd embed the same query twice, once for context detection and once for retrieving similar files.
* The cache does not key on the embeddings model, so it could return incorrect results when the model changes. So it'd be nice to just remove this.

File: enterprise/internal/embeddings/index_storage.go [hunk 2]
```diff
@@ -199,3 +206,73 @@ func decodeRepoEmbeddingIndex(dec *gob.Decoder) (*RepoEmbeddingIndex, error) {
...
+func (d *indexDecoder) close() {
+	d.file.Close()
+}
+
+const embeddingsChunkSize = 10_000
+
+// indexEncoder is a specialized encoder for repo embedding indexes. Instead of GOB-encoding
+// the entire RepoEmbeddingIndex, we encode each field separately, and we encode the embeddings array by chunks.
+// This way we avoid allocating a separate very large slice for the embeddings.
+type indexEncoder struct {
```
Comment (jtibshirani) [inline enterprise/internal/embeddings/index_storage.go:219]: It wasn't really necessary to introduce this struct, I just liked the symmetry with `indexDecoder`.

File: enterprise/internal/embeddings/types.go [hunk 3]
```diff
@@ -67,16 +67,21 @@ type RepoEmbeddingRowMetadata struct {
...
+	CodeIndex       EmbeddingIndex
+	TextIndex       EmbeddingIndex
 }
 
 func (i *RepoEmbeddingIndex) EstimateSize() int64 {
 	return i.CodeIndex.EstimateSize() + i.TextIndex.EstimateSize()
 }
 
+func (i *RepoEmbeddingIndex) IsModelCompatible(model string) bool {
+	return i.EmbeddingsModel == "" || i.EmbeddingsModel == model
```
Comment (eseliger) [inline enterprise/internal/embeddings/types.go:82]: quick drive-by comment: the models aren't provider specific right now, correct? Should we add the provider as a prefix to the model? Otherwise, `sourcegraph` provider with an equally named model will match `openai`. 
Also, we should probably special case `openai`s default model and `sourcegraph`s default model, since those are the same and will make migration from openai to sourcegraph easier.

Comment (jtibshirani) [reply enterprise/internal/embeddings/types.go:82]: The PR currently assumes the 'model' value in config uniquely identifies a model. It sounds like 'provider' + 'model' is actually the correct identifier -- I will switch to that. I didn't include 'dimensions' because I think of that as part of the model definition, we can avoid having two models with the same name and different dimensions.

File: enterprise/internal/embeddings/index_storage.go (hunk 2 above)
Comment (stefanhengl) [reply enterprise/internal/embeddings/index_storage.go:219]: nit: same comment as above, this could be just `encoder` and for symmetry we could also have a constructor ;-)

File: enterprise/internal/embeddings/index_storage.go [hunk 4]
```diff
@@ -199,3 +206,73 @@ func decodeRepoEmbeddingIndex(dec *gob.Decoder) (*RepoEmbeddingIndex, error) {
...
+	enc           *gob.Encoder
+	formatVersion IndexFormatVersion
+	chunkSize     int
+}
+
+func (e *indexEncoder) encodeIndex(rei *RepoEmbeddingIndex) error {
+	// Always encode index format version first, as part of 'file header'
+	if err := e.enc.Encode(e.formatVersion); err != nil {
```
Comment (stefanhengl) [inline enterprise/internal/embeddings/index_storage.go:227]: I think this is missing from the decoder

File: enterprise/internal/embeddings/index_storage.go [hunk 5]
```diff
@@ -13,6 +13,16 @@ import (
 	"github.com/sourcegraph/sourcegraph/lib/errors"
 )
 
+// IndexFormatVersion is a version number that's increased every time the on-disk index format is
```
Comment (stefanhengl) [inline enterprise/internal/embeddings/index_storage.go:16]: This comment should be for `CurrentFormatVersion`, shouldn't it?

File: enterprise/internal/embeddings/index_storage.go [hunk 6]
```diff
@@ -199,3 +206,73 @@ func decodeRepoEmbeddingIndex(dec *gob.Decoder) (*RepoEmbeddingIndex, error) {
...
+
+	if err := e.enc.Encode(rei.RepoName); err != nil {
+		return err
+	}
+
+	if err := e.enc.Encode(rei.Revision); err != nil {
+		return err
+	}
+
+	if e.formatVersion >= EmbeddingModelVersion {
```
Comment (stefanhengl) [inline enterprise/internal/embeddings/index_storage.go:239]: Does this ever evaluate to false?

File: enterprise/cmd/embeddings/shared/search.go [hunk 7]
```diff
@@ -55,6 +55,12 @@ func searchRepoEmbeddingIndexes(
 			return nil, errors.Wrapf(err, "getting repo embedding index for repo %q", repoName)
 		}
 
+		if !embeddingIndex.IsModelCompatible(queryModel) {
+			return nil, errors.Newf("embeddings model in config (%s) does not match the embeddings model for the"+
+				" index (%s). Embedding index for repo %q must be reindexed with the new model",
+				embeddingIndex.EmbeddingsModel, queryModel, repoName)
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply enterprise/cmd/embeddings/shared/search.go:61]: Oops, thanks!

File: enterprise/internal/embeddings/index_storage.go (hunk 5 above)
Comment (jtibshirani) [reply enterprise/internal/embeddings/index_storage.go:16]: I intended it for `IndexFormatVersion`. I guess you find the explanation/ wording to be confusing, I'll try to improve it.

File: enterprise/internal/embeddings/index_storage.go [hunk 8]
```diff
@@ -109,88 +124,80 @@ func DownloadRepoEmbeddingIndex(ctx context.Context, uploadStore uploadstore.Sto
...
-	}
+type indexDecoder struct {
+	file          io.ReadCloser
+	dec           *gob.Decoder
+	formatVersion IndexFormatVersion
+}
 
-	if err := enc.Encode(rei.Revision); err != nil {
-		return err
+func createIndexDecoder(ctx context.Context, uploadStore uploadstore.Store, key string) (*indexDecoder, error) {
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply enterprise/internal/embeddings/index_storage.go:133]: Thanks for the pointers on idiomatic naming here.

File: enterprise/internal/embeddings/index_storage.go (hunk 4 above)
Comment (jtibshirani) [reply enterprise/internal/embeddings/index_storage.go:227]: We decode the format version in the constructor `newDecoder`. It turned out much cleaner to do it this way than to put all that logic inside the `decode` method.

File: enterprise/internal/embeddings/index_storage.go (hunk 6 above)
Comment (jtibshirani) [reply enterprise/internal/embeddings/index_storage.go:239]: Not in production usage -- we could technically not even have a `formatVersion` parameter in the encoder. But it is really helpful to have this for testing, since it lets us write unit tests for indexes with different versions (we do this now in `index_storage_test.go`). 

I'll add a comment explaining it's only useful for tests.

**Comment by jtibshirani** [comment]
@stefanhengl this is ready for another look, I tried to respond to all questions/ comments.

File: enterprise/internal/embeddings/index_storage.go [hunk 9]
```diff
@@ -13,6 +13,16 @@ import (
...
 )
 
+// IndexFormatVersion is a version number that's increased every time the on-disk index format is
+// changed. Make sure to update CurrentFormatVersion whenever you add a new version.
+type IndexFormatVersion int
+
+const CurrentFormatVersion = EmbeddingModelVersion
+const (
+	InitialVersion        IndexFormatVersion = iota
+	EmbeddingModelVersion                    // Added the model name used to create embeddings
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply enterprise/internal/embeddings/index_storage.go]: No, sorry. Now it makes sense. I was reading this as a generic placeholder. Never mind my initial comment.

**Review by stefanhengl — APPROVED** [review]
LGTM!

File: enterprise/internal/embeddings/index_storage.go [hunk 10]
```diff
@@ -109,88 +120,80 @@ func DownloadRepoEmbeddingIndex(ctx context.Context, uploadStore uploadstore.Sto
...
+		if err != nil {
+			return nil, err
 		}
+		dec = gob.NewDecoder(f)
+		return &decoder{f, dec, InitialVersion}, nil
 	}
 
-	return nil
+	if formatVersion > CurrentFormatVersion {
+		return nil, errors.Newf("unrecognized index format version: %d", formatVersion)
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply enterprise/internal/embeddings/index_storage.go:153]: I couldn't figure out how to have a sentinel error that accepts a formatting parameter 🤔

**Comment by mucles** [comment]
@cla-bot check
